src/diffusers/training_utils.py

```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
     """
     Exponential Moving Average of models weights
     """
 
     def __init__(
         self,
-        model,
-        update_after_step=0,
-        inv_gamma=1.0,
-        power=2 / 3,
-        min_value=0.0,
-        max_value=0.9999,
-        device=None,
+        parameters: Iterable[torch.nn.Parameter],
+        decay: float = 0.9999,
+        update_after_step: int = 0,
+        use_ema_warmup: bool = False,
+        inv_gamma: Union[float, int] = 1.0,
+        power: Union[float, int] = 2 / 3,
     ):
         """
+        Args:
+            parameters (Iterable[torch.nn.Parameter]): The parameters to track.
+            decay (float): The decay factor for the exponential moving average.
+            update_after_step (int): The number of steps to wait before starting to update the EMA weights.
+            use_ema_warmup (bool): Whether to use EMA warmup.
+            inv_gamma (float):
+                Inverse multiplicative factor of EMA warmup. Default: 1. Only used if `use_ema_warmup` is True.
+            power (float): Exponential factor of EMA warmup. Default: 2/3. Only used if `use_ema_warmup` is True.
+
         @crowsonkb's notes on EMA Warmup:
             If gamma=1 and power=1, implements a simple average. gamma=1, power=2/3 are good values for models you plan
             to train for a million or more steps (reaches decay factor 0.999 at 31.6K steps, 0.9999 at 1M steps),
             gamma=1, power=3/4 for models you plan to train for less (reaches decay factor 0.999 at 10K steps, 0.9999
             at 215.4k steps).
-
-        Args:
-            inv_gamma (float): Inverse multiplicative factor of EMA warmup. Default: 1.
-            power (float): Exponential factor of EMA warmup. Default: 2/3.
-            min_value (float): The minimum EMA decay rate. Default: 0.
         """
+        parameters = list(parameters)
+        self.shadow_params = [p.clone().detach() for p in parameters]
 
-        self.averaged_model = copy.deepcopy(model).eval()
-        self.averaged_model.requires_grad_(False)
+        self.collected_params = None
 
+        self.decay = decay
         self.update_after_step = update_after_step
+        self.use_ema_warmup = use_ema_warmup
         self.inv_gamma = inv_gamma
         self.power = power
-        self.min_value = min_value
-        self.max_value = max_value
-
-        if device is not None:
-            self.averaged_model = self.averaged_model.to(device=device)
-
-        self.decay = 0.0
         self.optimization_step = 0
 
     def get_decay(self, optimization_step):
         """
         Compute the decay factor for the exponential moving average.
         """
         step = max(0, optimization_step - self.update_after_step - 1)
-        value = 1 - (1 + step / self.inv_gamma) ** -self.power
+
+        if self.use_ema_warmup:
+            cur_decay_value = 1 - (1 + step / self.inv_gamma) ** -self.power
+        else:
+            cur_decay_value = (1 + step) / (10 + step)
```

Comment:
this is the default which is used for SD.

Reply:
Are the two previous implementations equivalent to this new one or are we overriding one of them? (Sorry, it's not clear to me from the diff).

Reply:
if `use_ema_warmup` is `True` then it'll compute the decay as per the previous `EMAModel`, if it's `False` (which is the default, it will compute it the same way we were using for SD. Does that makes sense ?